File: frb_example/pure_dart/lib/src/rust/api/rust_opaque_sync.dart
// ignore_for_file: invalid_use_of_internal_member, unused_import

import '../frb_generated.dart';
import 'package:flutter_rust_bridge/flutter_rust_bridge_for_generated.dart';
import 'rust_opaque.dart';

String syncRunOpaque({required NonSendHideData opaque, dynamic hint}) =>
    RustLib.instance.api.syncRunOpaque(opaque: opaque, hint: hint);

/// Structure for testing the sync-mode RustOpaque code generator.
/// FrbOpaqueSyncReturn must be only return type.
/// FrbOpaqueSyncReturn must be without wrapper like Option<> Vec<> etc.
Future<FrbOpaqueSyncReturn> frbSyncGeneratorTest({dynamic hint}) =>
    RustLib.instance.api.frbSyncGeneratorTest(hint: hint);

@sealed
class FrbOpaqueSyncReturn extends FrbOpaque {
  FrbOpaqueSyncReturn.fromRaw(int ptr, int size) : super.unsafe(ptr, size);

  @override
  OpaqueDropFnType get dropFn =>
      RustLib.instance.api.dropOpaqueFrbOpaqueSyncReturn;

  @override
  OpaqueShareFnType get shareFn =>
      RustLib.instance.api.shareOpaqueFrbOpaqueSyncReturn;

  @override
  OpaqueTypeFinalizer get staticFinalizer =>
      RustLib.instance.api.frbOpaqueSyncReturnFinalizer;
}

File: frb_example/pure_dart/rust/src/api/rust_opaque_sync.rs
// FRB_INTERNAL_GENERATOR: {"forbiddenDuplicatorModes": ["sync", "rustAsync", "sync sse", "rustAsync sse", "sync moi", "rustAsync moi", "sync sse moi", "rustAsync sse moi"], "enableAll": true}

use super::rust_opaque::NonCloneDataTwinNormal;
pub use crate::auxiliary::sample_types::{HideDataRaw, NonCloneDataRaw};
#[allow(unused_imports)]
use crate::frb_generated::RustOpaque;
use anyhow::Result;
use flutter_rust_bridge::frb;

#[allow(dead_code)]
pub struct HideDataAnotherTwinNormal(HideDataRaw);

/// Structure for testing the SyncReturn<RustOpaque> code generator.
/// FrbOpaqueSyncReturn must be only return type.
/// FrbOpaqueSyncReturn must should be without wrapper like Option<> Vec<> etc.
pub struct FrbOpaqueSyncReturnTwinNormal;

#[frb(sync)]
pub fn sync_option_rust_opaque_twin_normal() -> Result<Option<RustOpaque<HideDataAnotherTwinNormal>>>
{
    Ok(Some(RustOpaque::new(HideDataAnotherTwinNormal(
        HideDataRaw::new(),
    ))))
}

#[frb(sync)]
pub fn sync_create_opaque_twin_normal() -> RustOpaque<HideDataAnotherTwinNormal> {
    RustOpaque::new(HideDataAnotherTwinNormal(HideDataRaw::new()))
}

#[frb(sync)]
pub fn sync_create_non_clone_twin_normal() -> RustOpaque<NonCloneDataTwinNormal> {
    RustOpaque::new(NonCloneDataTwinNormal(NonCloneDataRaw::new()))
}

// OpaqueSyncStruct does not implement Send trait.
//
// pub fn run_opaque_twin_normal(opaque: Opaque<OpaqueSyncStruct>) -> String {
//     data.0.hide_data()
// }

/// Structure for testing the sync-mode RustOpaque code generator.
/// FrbOpaqueSyncReturn must be only return type.
/// FrbOpaqueSyncReturn must be without wrapper like Option<> Vec<> etc.
#[frb(sync)]
pub fn frb_sync_generator_test_twin_normal() -> RustOpaque<FrbOpaqueSyncReturnTwinNormal> {
    RustOpaque::new(FrbOpaqueSyncReturnTwinNormal)
}
